packages/headless/src/controllers/insight/attached-results/headless-attached-results.test.ts

```diff
@@ -0,0 +1,227 @@
+    expect(controller).toBeTruthy();
+  });
+
+  it('it adds the correct reducers to the engine', () => {
+    expect(engine.addReducers).toHaveBeenCalledWith({
+      configuration,
+      attachedResults: attachedResultsReducer,
+    });
+  });
+
+  it('exposes a subscribe method', () => {
+    expect(controller.subscribe).toBeTruthy();
+  });
+
+  describe('#isAttached', () => {
+    describe('when no results are attached', () => {
+      it('calling #isAttached should return false', () => {
+        expect(controller.isAttached(payload)).toBe(false);
+      });
+    });
+
+    describe('when the result is attached', () => {
+      const testPermanentId = 'testPermanentId';
+      const testCaseId = '12345';
+
+      beforeEach(() => {
+        defaultOptions = {
+          caseId: testCaseId,
+        };
+        initAttachedResults({...defaultOptions});
+      });
+
+      it('#isAttached should return true', () => {
+        const mockAttachedResult = createMockAttachedResult({
+          caseId: testCaseId,
+          permanentId: testPermanentId,
+          uriHash: undefined,
+        });
+        engine.state.attachedResults!.results.push(mockAttachedResult);
+
+        expect(controller.isAttached(payload)).toBe(true);
+      });
+
+      it('#isAttached should return true even when multiple results are attached', () => {
+        [testPermanentId, 'foo', 'bar'].forEach((permId) => {
+          const mockAttachedResult = createMockAttachedResult({
+            caseId: testCaseId,
+            permanentId: permId,
+            uriHash: undefined,
+          });
+          engine.state.attachedResults!.results.push(mockAttachedResult);
+        });
+
+        expect(controller.isAttached(payload)).toBe(true);
+      });
+    });
+
+    describe('when the result is not attached', () => {
+      const testCaseId = '12345';
+
+      beforeEach(() => {
+        defaultOptions = {
+          caseId: testCaseId,
+        };
+        initAttachedResults({...defaultOptions});
+      });
+
+      it('#isAttached should return false', () => {
+        ['foo', 'bar'].forEach((permId) => {
+          const mockAttachedResult = createMockAttachedResult({
+            caseId: testCaseId,
+            permanentId: permId,
+            uriHash: undefined,
+          });
+          engine.state.attachedResults!.results.push(mockAttachedResult);
+        });
+
+        expect(controller.isAttached(payload)).toBe(false);
+      });
+    });
+  });
+
+  describe('#attach', () => {
+    const testPermanentId = 'testPermanentId';
+    const testUriHash = 'testUriHash';
+    const testCaseId = '12345';
+    const testResult = buildMockResult({
+      clickUri: 'foo.bar',
+      title: 'test result',
+      raw: {
+        permanentid: testPermanentId,
+        urihash: testUriHash,
+      },
+    });
+
+    beforeEach(() => {
+      defaultOptions = {
+        caseId: testCaseId,
+      };
+      initAttachedResults({...defaultOptions});
+    });
```

Comment:
Why this beforeEach is necessary while it does the same logic of the top level before each?